shuup/product_discounts/module.py

```diff
@@ -0,0 +1,90 @@
+from django.utils.translation import ugettext_lazy as _
+
+from shuup.core.models import Category, ShopProduct
+from shuup.core.pricing import DiscountModule
+
+from .models import ProductDiscount
+
+
+class ProductDiscountModule(DiscountModule):
+    identifier = "product_discounts"
+    name = _("Product Discount Module")
+
+    def discount_price(self, context, product, price_info):
+        shop = context.shop
+        product_id = product if isinstance(product, six.integer_types) else product.pk
+
+        category_ids = Category.objects.filter(shop_products__product_id=product_id).values_list("id", flat=True)
+        group_ids = context.customer.groups.values_list("id", flat=True)
+
+        # Product condition is always applied
+        condition_query = (Q(product__isnull=True) | Q(product_id=product_id))
+
+        # Apply category conditions
+        condition_query &= (
+            Q(category__isnull=True) |
+            (Q(exclude_selected_category=False) & Q(category__id__in=category_ids)) |
+            (Q(exclude_selected_category=True) & ~Q(category__id__in=category_ids))
+        )
+
+        # Apply contact conditions
+        condition_query &= (Q(contact__isnull=True) | Q(contact_id=context.customer.pk))
+
+        # Apply contact group conditions
+        condition_query &= (
+            Q(contact_group__isnull=True) |
+            (Q(exclude_selected_contact_group=False) & Q(contact_group__id__in=group_ids)) |
+            (Q(exclude_selected_contact_group=True) & ~Q(contact_group__id__in=group_ids))
+        )
+
+        # Get all possible discounts for the current product and context
+        potential_discounts = ProductDiscount.objects.available(shop).filter(
+            condition_query).values_list("discounted_price_value", "discount_amount_value", "discount_percentage")
+
+        discounted_prices = []
+        for discounted_price_value, discount_amount_value, discount_percentage in potential_discounts:
+            if discounted_price_value:  # Applies the new product price per item
+                discounted_prices.append(
+                    max(
+                        shop.create_price(discounted_price_value) * price_info.quantity,
+                        shop.create_price(0)
+                    )
+                )
+
+            if discount_amount_value:  # Discount amount value per item
+                discounted_prices.append(
+                    max(
+                        price_info.price - shop.create_price(discount_amount_value) * price_info.quantity,
+                        shop.create_price(0)
+                    )
+                )
+
+            if discount_percentage:  # Discount percentage per item
+                discounted_prices.append(
+                    max(
+                        price_info.price - price_info.price * discount_percentage,
+                        shop.create_price(0)
+                    )
+                )
+
+        if len(discounted_prices):
```

Comment:
for minor speed up, maybe use `if discounted_prices` here to prevent iteration :)